### PR TITLE
Custom pronouns & presets

### PR DESCRIPTION
**Pull request by DeadGirlRisen**
Alright, with this we should have a custom pronouns screen so you can customize which pronouns a cat goes by, and can add in your own pronouns.  We've worked hard on this, and hopefully now this time we got everything ready.

### CODE REVIEW

**Comment by keyraven** [comment]
I will do a full review of this later. From a quick functionality check, the game currently crashes when pressing the"x" button in the "Add Pronoun" dialogue.

File: resources/dicts/pronouns.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+  "default_pronouns": [
+    {
+      "ID": "Neutral",
```
Comment (keyraven) [inline resources/dicts/pronouns.json]: As I've mentioned in previous reviews, I do not see the need for the "ID" field for pronouns. You can check for strict equality of the entire pronoun dictionary for deletion/identification. The "ID" field is not shown to the player, and adds further complexity.

**Comment by DeadGirlRisen** [comment]
> I will do a full review of this later. From a quick functionality check, the game currently crashes when pressing the"x" button in the "Add Pronoun" dialogue.

Thats odd, I'm not having that happen myself.

Wait, it happens after you save a pronoun oddly enough, but not if you don't add a pronoun. I'll take a look, somehow in the testing that bug was not encountered before? Or it might be due to the update from bingle. Thanks for finding it, I'll see what I can do.

**Comment by DeadGirlRisen** [comment]
> I will do a full review of this later. From a quick functionality check, the game currently crashes when pressing the"x" button in the "Add Pronoun" dialogue.

Alright, it looks like I fixed it from a quick test on my end, it was just a matter of one of the bits of code forgot to be changed over with the new changes bingle made.

File: resources/dicts/events/ceremonies/ceremony-master.json [hunk 2]
```diff
@@ -2056,7 +2056,7 @@
       "general_backstory",
       "all_traits"
     ],
-    "l_n calls the Clan to a meeting and declares (old_name) to be a warrior. {PRONOUN/m_c/subject/CAP} {VERB/m_c/are/is} now called m_c and are celebrated for {PRONOUN/m_c/poss} r_h."
+    "l_n calls the Clan to a meeting and declares (old_name) to be a warrior. {PRONOUN/m_c/subject/CAP} {VERB/m_c/are/is} now called m_c and {VERB/m_c/are/is} celebrated for {PRONOUN/m_c/poss} r_h."
```
Comment (keyraven) [inline resources/dicts/events/ceremonies/ceremony-master.json:2059]: This is a large pull request, so it would make it easier to review if these non-related pronoun tag corrections were put in a separate pull request. At this point, it may be too much work. But in the future, it would be best to keep large feature addition PRs focused on the feature added.

File: scripts/cat/cats.py [hunk 3]
```diff
@@ -280,66 +283,91 @@ def __init__(self,
...
-            elif self.genderalign in ["male", "trans male"]:
-                self.pronouns = [self.default_pronouns[2].copy()]"""
-
-            # APPEARANCE
-            self.pelt = Pelt.generate_new_pelt(self.gender, [Cat.fetch_cat(i) for i in (self.parent1, self.parent2) if i], self.age)
+            #newborns can't be trans, sorry babies
+            if self.age in ['kitten', 'newborn']:
+                trans_chance = 0
+                nb_chance = 0
+            if game:
```
Comment (keyraven) [inline scripts/cat/cats.py]: If, for whatever reason, "game" is falsy, a cat will generate with an empty pronoun list, which will crash the game when text replacement needs to happen. Rather than putting the entire pronoun determination inside the "if game" check, set a local variable with the value of "they them default", then proceed with pronoun determination.

File: scripts/cat/cats.py [hunk 4]
```diff
@@ -280,66 +283,91 @@ def __init__(self,
...
-            elif self.age in ['senior adult']:
-                self.experience = randint(Cat.experience_levels_range["competent"][0],
-                                          Cat.experience_levels_range["expert"][1])
-            elif self.age in ['senior']:
-                self.experience = randint(Cat.experience_levels_range["competent"][0],
-                                          Cat.experience_levels_range["master"][1])
-            else:
-                self.experience = 0
+        #Personality
+        self.personality = Personality(kit_trait=self.is_baby())
```
Comment (keyraven) [inline scripts/cat/cats.py]: This section should remain indented.  It needs to remain under the current if-statement.

File: scripts/cat/cats.py [hunk 5]
```diff
@@ -1890,7 +1918,7 @@ def contact_with_ill_cat(self, cat: Cat):
             if self.is_injured():
                 for y in self.injuries:
                     illness_infect = list(
-                        filter(lambda ill: ill["name"] == illness_name, self.injuries[y]["illness_infectiousness"]))
+                        filter(lambda ill: ill["ID"] == illness_name, self.injuries[y]["illness_infectiousness"]))
```
Comment (keyraven) [inline scripts/cat/cats.py]: What is the purpose of this change? It doesn't seem to have anything to do w/ pronouns. Is this an unrelated bug fix?

File: scripts/cat/cats.py [hunk 6]
```diff
@@ -280,66 +283,91 @@ def __init__(self,
...
+                    #print("Game is asigning they/them pronouns")
+                    if nb_chance == 1:
+                        self.genderalign = "nonbinary"
+                    elif trans_chance == 1:
+                        if self.gender == "female":
+                            self.genderalign = "trans male"
+                        else:
+                            self.genderalign = "trans female"
+                #assigning pronouns to afab cats            
+                elif self.gender == "female":
```
Comment (keyraven) [inline scripts/cat/cats.py]: The section for AFAB and AMAB cats is also identical. You could improve readability by condensing into one section.  Or (even better), separating out gender and pronoun assignment so that gender is determined first, then pronouns.

File: scripts/cat/cats.py [hunk 7]
```diff
@@ -280,66 +283,91 @@ def __init__(self,
...
+                                      Cat.experience_levels_range["expert"][1])
+        elif self.age in ['senior']:
+            self.experience = randint(Cat.experience_levels_range["competent"][0],
+                                      Cat.experience_levels_range["master"][1])
+        else:
+            self.experience = 0
                 
-            if not skill_dict:
-                self.skills = CatSkills.generate_new_catskills(self.status, self.moons)
+        if not skill_dict:
```
Comment (keyraven) [inline scripts/cat/cats.py]: Same here - re-indent this

File: scripts/clan.py [hunk 8]
```diff
@@ -457,6 +463,9 @@ def save_clan(self):
             clan_data["med_cat"] = None
         clan_data["med_cat_number"] = self.med_cat_number
         clan_data["med_cat_predecessors"] = self.med_cat_predecessors
+        
+        if self.custom_pronouns:
+            clan_data["custom_pronouns"] = self.custom_pronouns
```
Comment (keyraven) [inline scripts/clan.py]: There is no need for lines 467 and 468. self.custom_pronouns is already assigned to the key "custom_pronouns" on line 439.

File: scripts/clan.py [hunk 9]
```diff
@@ -746,6 +755,11 @@ def load_clan_json(self):
         game.clan.deputy_predecessors = clan_data["deputy_predecessors"]
         game.clan.med_cat_predecessors = clan_data["med_cat_predecessors"]
         game.clan.med_cat_number = clan_data["med_cat_number"]
+        if "custom_pronouns" in clan_data.keys():
+            if clan_data["custom_pronouns"]:
+                game.clan.custom_pronouns = clan_data["custom_pronouns"]
+        else:
+            game.clan.custom_pronouns = []
```
Comment (keyraven) [inline scripts/clan.py]: I would remove this "else" statement, since custom_pronouns is already set as a empty list in the clan constructor.

File: scripts/game_structure/load_cat.py [hunk 10]
```diff
@@ -121,6 +121,8 @@ def json_load():
             new_cat.adoptive_parents = cat["adoptive_parents"] if "adoptive_parents" in cat else []
             
             new_cat.genderalign = cat["gender_align"]
+            # new_cat.pronouns = cat["pronouns"]
+            new_cat.pronouns = cat["pronouns"] if "pronouns" in cat else [new_cat.default_pronouns[0].copy()]
```
Comment (keyraven) [inline scripts/game_structure/load_cat.py:125]: Note -- cats from old saves will be assigned they/them pronouns irrregardless of the setting. This is probably fine, but important to note.

File: scripts/game_structure/load_cat.py [hunk 11]
```diff
@@ -182,6 +184,7 @@ def json_load():
                 )
 
             # new_cat.pronouns = cat["pronouns"] if "pronouns" in cat else [new_cat.default_pronouns[0].copy()]
+            # new_cat.pronouns = cat.get("pronouns", new_cat.default_pronouns[0].copy())
```
Comment (keyraven) [inline scripts/game_structure/load_cat.py]: If these lines are no longer necessary, and replaced by line 125, could you delete them to keep the loading code tidy? Same with line 124.

File: scripts/screens/ChangeGenderScreen.py [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+                        manager=MANAGER
+                    )
+
+            elif event.ui_element == self.buttons["add_pronouns"]:
+                PronounCreation(self.the_cat)
+                self.previous_cat_button.disable()
+                self.next_cat_button.disable()
+                self.back_button.disable()
+
+            else:
```
Comment (keyraven) [inline scripts/screens/ChangeGenderScreen.py]: Having a "else" statement for button event handling is confusing, and could possibly cause issues down the line.  It took me a while to figure out what this section of code was for.   If would be clearer to use additional elif statements, rather than a blanket "else" statement for button clicks.

File: scripts/screens/ChangeGenderScreen.py [hunk 13]
```diff
@@ -0,0 +1,631 @@
...
+                with open('resources/dicts/pronouns.json', 'r', encoding='utf-8') as file:
+                    pronouns_data = ujson.load(file)
+
+                for pronounset in pronouns_data["default_pronouns"]:
+                    add_button_id = f"add_button_{pronounset['ID']}" + "preset"
+                    if event.ui_element == self.buttons.get(add_button_id):
```
Comment (keyraven) [inline scripts/screens/ChangeGenderScreen.py]: Rather than looping button keys, I would suggest using something similar to CatButton or UISpriteButton, which adds an additional data field to the button object, which can then be extracted from the button event. In fact, you would probably add an object_id field to the CatButton constructor and use that class directly.  This would also help remove the need for the 'ID' value for pronoun dictionaries.

File: scripts/screens/ChangeGenderScreen.py [hunk 14]
```diff
@@ -0,0 +1,631 @@
...
+                        game.clan.custom_pronouns.remove(pronounset)
+                        self.update_selected_cat()
+                for pronounset in self.the_cat.pronouns:
+                    remove_button_id = f"remove_button_{pronounset['ID']}"
+                    if event.ui_element == self.removalbuttons.get(remove_button_id):
+                        self.the_cat.pronouns.remove(pronounset)
+                        self.update_selected_cat()
+                        break
+
+        elif event.type == pygame.KEYDOWN and game.settings['keybinds']:
```
Comment (keyraven) [inline scripts/screens/ChangeGenderScreen.py]: Probably best to remove keybinds for now. They are a pain and a half. As this code is currently set-up, it would allow game.switches['cat'] to be set to an invalid cat ID.

File: scripts/screens/ProfileScreen.py [hunk 15]
```diff
@@ -263,12 +263,12 @@ def handle_tab_events(self, event):
             if event.ui_element == self.change_name_button:
                 ChangeCatName(self.the_cat)
             elif event.ui_element == self.specify_gender_button:
-                SpecifyCatGender(self.the_cat)
-                '''if self.the_cat.genderalign in ["female", "trans female"]:
+                self.change_screen("change gender screen")
+                if self.the_cat.genderalign in ["female", "trans female"]:
```
Comment (keyraven) [inline scripts/screens/ProfileScreen.py]: This is currently the origin of a bug,where going to the cat's gender specify page will re-set their pronouns to correspond with their gender, clearing any previously set pronouns and ignoring the they/them default setting. Remove lines 266-271 to fix this issue.

File: scripts/utility.py [hunk 16]
```diff
@@ -1295,7 +1295,7 @@ def generate_sprite(cat, life_state=None, scars_hidden=False, acc_hidden=False,
         # draw eyes & scars1
         eyes = sprites.sprites['eyes' + cat.pelt.eye_colour + cat_sprite].copy()
         if cat.pelt.eye_colour2 != None:
-            eyes.blit(sprites.sprites['eyes2' + cat.pelt.eye_colour2 + cat_sprite], (0, 0))
+            eyes.blit(sprites.sprites["eyes2" + cat.pelt.eye_colour2 + cat_sprite], (0, 0))
```
Comment (keyraven) [inline scripts/utility.py]: What is the purpose of this change? Is this to fix an unrelated bug?  Same with the new "eye2color.png" file.

**Review by keyraven — CHANGES_REQUESTED** [review]
It's looking much better! Mostly works great. I found a few issues, and had a few comments on the ChangeGenderScreen code.

File: resources/dicts/events/ceremonies/ceremony-master.json (hunk 2 above)
Comment (DeadGirlRisen) [reply resources/dicts/events/ceremonies/ceremony-master.json:2059]: Will do that in the future 👍

File: resources/dicts/pronouns.json (hunk 1 above)
Comment (DeadGirlRisen) [reply resources/dicts/pronouns.json]: Bingle and I would like help with this 💜

File: scripts/cat/cats.py (hunk 3 above)
Comment (DeadGirlRisen) [reply scripts/cat/cats.py]: Marking as fixed

File: scripts/cat/cats.py (hunk 4 above)
Comment (DeadGirlRisen) [reply scripts/cat/cats.py]: I believe I indented it correctly now :D

File: scripts/cat/cats.py (hunk 5 above)
Comment (DeadGirlRisen) [reply scripts/cat/cats.py]: My mistake has been fixed.

File: scripts/cat/cats.py (hunk 6 above)
Comment (DeadGirlRisen) [reply scripts/cat/cats.py]: Fixed! I believe I streamlined it, and it looked like it worked in my testing.

File: scripts/cat/cats.py (hunk 7 above)
Comment (DeadGirlRisen) [reply scripts/cat/cats.py]: Reindented! :D

File: scripts/clan.py (hunk 8 above)
Comment (DeadGirlRisen) [reply scripts/clan.py]: Deleted 👍

File: scripts/clan.py (hunk 9 above)
Comment (DeadGirlRisen) [reply scripts/clan.py]: fixed :D

File: scripts/game_structure/load_cat.py (hunk 11 above)
Comment (DeadGirlRisen) [reply scripts/game_structure/load_cat.py]: Done :D

File: scripts/screens/ProfileScreen.py (hunk 15 above)
Comment (DeadGirlRisen) [reply scripts/screens/ProfileScreen.py]: Whoops, good catch, I remember thinking to myself a few months back I need to fix that, guess I forgot XD

File: scripts/utility.py (hunk 16 above)
Comment (DeadGirlRisen) [reply scripts/utility.py]: Just a mistake when I was working on a different branch and hadn't properly changed my branches over. It should be fixed.

File: scripts/screens/ChangeGenderScreen.py (hunk 12 above)
Comment (DeadGirlRisen) [reply scripts/screens/ChangeGenderScreen.py]: I believe I fixed this, though I think when you fix the way the pronouns are identified this area of code might be changed as well.

File: scripts/screens/ChangeGenderScreen.py (hunk 13 above)
Comment (DeadGirlRisen) [reply scripts/screens/ChangeGenderScreen.py]: I believe this is part of what you're helping with, so I leave it in your capable hands  :D

File: scripts/screens/ChangeGenderScreen.py (hunk 14 above)
Comment (DeadGirlRisen) [reply scripts/screens/ChangeGenderScreen.py]: Removed :D

**Review by keyraven — APPROVED** [review]
Looking Good!
